### PR TITLE
Raise error when affecting tuple moved into different segment.

### PR DESCRIPTION
**Pull request by kainwen**
Greenplum supports update on distkeys of a table. It is implemented
by split update: delete the old tuple in the original segment and
insert the new tuple in another segment (maybe the same as original).
So the two operations (insert and delete) happan may happen in different
segments. It is not the same as heap_update which happens in a single
segment. This makes it hard to follow the link to find the new tuple.
The transaction will raise an error if it finds out that the XID it
waited for is deleted by a split update's delete action.

The skill used here is the same idea as Postgres's commmit f16241bef7cc2
to handle update statement on paritition keys. The row deleted as part
of a cross partition update is marked by pointing it's t_ctid to an
invalid block, instead of self as a normal update would.

Fixes github issue: https://github.com/greenplum-db/gpdb/issues/8919

---------------------------------------

The idea is based on the upstream commit: https://github.com/postgres/postgres/commit/71bff60e23de2f827a10e50dde8

And the discussion thread of gpdb-dev mailing list is https://groups.google.com/a/greenplum.org/forum/#!topic/gpdb-dev/Yg5ypzj-fqg

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/include/storage/itemptr.h [hunk 1]
```diff
@@ -51,6 +51,22 @@ ItemPointerData;
...
+
+/*
+ * Greenplum specific behavior:
+ * When a tuple is moved to a different segment by UPDATE on distkeys
+ * (greenplum implements this using split update techniques), the t_ctid
+ * of the deleted tuple is set to this magic value.
+ *
+ * NB: later version of Postgres use 0xfffc and 0xfffe.
+ */
+#define SplitUpdateOffsetNumber 0xffff
```
Comment (ashwinstar) [inline src/include/storage/itemptr.h]: Am I missing something, don't see these two values used anywhere?

**Comment by hlinnaka** [comment]
> I was thinking we would kind of literally cherry-pick the upstream commit and then modify the same to our needs, but just change it slightly to our needs. This seems we are taking the upstream logic but not the commit. Just wish to point this out and bring for discussion as I am fence for the same.
> 
> The macro names used in upstream are very specific and we wish to use the same in much more broader sense. So, having more better names for macros as per our usage in GPDB seems good, but then it poses issues for merge. What do folks feel?

Agreed, we should literally cherry-pick the upstream commit. Use the same variable names, same magic values, etc.  In particular, we should use the same values for `MovedPartitionsOffsetNumber` and `MovedPartitionsBlockNumber` as in upstream. They are stored on-disk, so if we pick different values, we're stuck with them forever. You can add a comment on them, explaining that in GPDB they're also used for cross-segment updates.

**Comment by kainwen** [comment]
@ashwinstar  @hlinnaka 
OK. I close this and will open a new pr soon.
